packages/use-query-params/src/index.js

```diff
@@ -0,0 +1,49 @@
+import { parse, stringify } from 'query-string'
+import { useCallback, useEffect, useState } from 'react'
+import { useHistory } from 'react-router-dom'
+
+const useQueryParams = () => {
+  const { location, replace } = useHistory()
+  const defaultFnUpdater = useCallback(
+    (currentQueryParams, updater) => ({
+      ...currentQueryParams,
+      ...updater,
+    }),
+    [],
+  )
+  const queryParams = parse(location.search)
+  const [state, setState] = useState(queryParams)
+
+  const setQueryParams = useCallback(
+    updater => {
+      const currentQueryParams = parse(location.search)
+
+      const newParams =
+        updater instanceof Function
+          ? updater(currentQueryParams)
+          : defaultFnUpdater(currentQueryParams, updater)
+
+      const stringifiedParams = stringify(newParams)
```

Comment:
I just noticed, another option instead if using a deep-equal to avoid useless query string updates would be to use a consistent sort in the stringify

For example:

```js
const stringifiedParams = stringify(newParams, { sort: (a, b) => order.indexOf(a) - order.indexOf(b) })

if (stringifiedParams === location.search) {
  return newParams
}
```